include/range/v3/utility/functional.hpp

```diff
@@ -458,30 +440,36 @@ namespace ranges
             {}
             template<typename... Args>
             auto operator()(Args&&...args)
-            RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
-            (
-                overloaded::compressed_pair::first(detail::forward<Args>(args)...)
-            )
+            RANGES_DECLTYPE_NOEXCEPT(
+                std::declval<function_type<First> &>()(
+                    detail::forward<Args>(args)...))
+            {
+                return this->first()(detail::forward<Args>(args)...);
+            }
             template<typename... Args>
             auto operator()(Args&&...args) const
-            RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
-            (
-                detail::as_const(overloaded::compressed_pair::first)(
-                    detail::forward<Args>(args)...)
-            )
+            RANGES_DECLTYPE_NOEXCEPT(
+                std::declval<function_type<First> const &>()(
+                    detail::forward<Args>(args)...))
+            {
+                return this->first()(detail::forward<Args>(args)...);
+            }
             template<typename... Args>
             auto operator()(Args&&...args)
-            RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
-            (
-                overloaded::compressed_pair::second(detail::forward<Args>(args)...)
-            )
+            RANGES_DECLTYPE_NOEXCEPT(
+                std::declval<overloaded<Rest...> &>()(
+                    detail::forward<Args>(args)...))
+            {
+                return this->second()(detail::forward<Args>(args)...);
+            }
             template<typename... Args>
             auto operator()(Args&&...args) const
-            RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
-            (
-                detail::as_const(overloaded::compressed_pair::second)(
-                    detail::forward<Args>(args)...)
-            )
+            RANGES_DECLTYPE_NOEXCEPT(
+                std::declval<overloaded<Rest...> const &>()(
+                    detail::forward<Args>(args)...))
+            {
+                return this->second()(detail::forward<Args>(args)...);
+            }
```

Comment:
GCC doesn't like `this` - and consequently member function accesses - in `noexcept` clauses [PR52869](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52869). 4.8 (and maybe 4.9?) don't like them in trailing return types, either.

Not sure why I used `this->member()` syntax here instead of `using base_type::member;`. It's probably fewer overall keystrokes, but inconsistent.